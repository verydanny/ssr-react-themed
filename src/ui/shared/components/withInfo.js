import { connect } from 'react-redux'
import {
  makeUnitAvailTextSelector,
  makePriceTextSelector,
  makeBedTextSelector,
  makeBathTextSelector,
  makeGetListing,
} from 'app/selectors/listingInfo'
import {
  getIsMapView,
  getIsListView,
} from 'app/selectors/page'
import { getIsDesktop } from 'app/selectors/request'

const makeMapStateToProps = (initialState, initialProps) => {
  const isMapView = getIsMapView(initialState)
  const isListView = getIsListView(initialState)
  const priceTextSelector = makePriceTextSelector()
  const bedTextSelector = makeBedTextSelector()
  const bathTextSelector = makeBathTextSelector()
  const isMobileMapView = isMapView && !getIsDesktop(initialState)
  const unitAvailTextSelector =
    makeUnitAvailTextSelector(isMobileMapView)
  const getListing = makeGetListing(initialProps.listingId)

  return (state, props) => {
    const listing = props.listing || getListing(state)
    return {
      ...listing,
      isMapView,
      isListView,
      priceText: priceTextSelector(listing),
      beds: bedTextSelector(listing),
      baths: bathTextSelector(listing),
      unitAvailText: unitAvailTextSelector(listing),
    }
  }
}

export default connect(makeMapStateToProps, {})
