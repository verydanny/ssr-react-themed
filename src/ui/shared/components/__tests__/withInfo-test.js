import React from 'react'
import { Provider } from 'react-redux'
import { mount } from 'enzyme'
import { createMockStore } from 'mocks'
import withInfo from '../withInfo'

const dummyComponent = <div data-tid="dummyComponent" />

const listing = {
  originalListingProperty: 'im available to children',
  listingId: 0,
  isPaid: true,
  isActive: true,
  aggregates: {
    prices: {
      low: 800,
      high: 2100,
    },
    totalAvailable: 42,
    beds: {
      low: 2,
      high: 4,
    },
    baths: {
      low: 1,
      high: 3,
    },
  },
}
const state = {
  page: {
    currentView: 'map',
  },
  request: {
    displaySize: 'small',
  },
  listings: {
    listings: [listing],
  },
}

const store = createMockStore(state)

describe('withInfo HOC', () => {
  it('decorates listing info', () => {
    const props = {
      listing,
      someOtherProp: 'im availabile to children',
    }
    const InfoDiv = withInfo(() => dummyComponent)
    const wrapper = <InfoDiv {...props} />
    expect(wrapper.props).toMatchSnapshot()
  })

  it('grabs state from redux', () => {
    const props = {
      listing,
      listingId: listing.listingId,
      someOtherProp: 'im availabile to children',
    }

    const InfoDiv = withInfo(() => dummyComponent)

    const wrapper = mount(
      <Provider store={store}>
        <InfoDiv {...props} />
      </Provider>
    )

    expect(wrapper.find(InfoDiv).childAt(0).prop('unitAvailText')).toEqual('10+ available')
    expect(wrapper.find(InfoDiv).childAt(0).prop('isMapView')).toEqual(true)
    expect(wrapper.find(InfoDiv).childAt(0).prop('isListView')).toEqual(false)
  })
})
