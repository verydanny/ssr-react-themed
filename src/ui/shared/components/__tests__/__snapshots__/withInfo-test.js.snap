// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withInfo HOC decorates listing info 1`] = `
Object {
  "listing": Object {
    "aggregates": Object {
      "baths": Object {
        "high": 3,
        "low": 1,
      },
      "beds": Object {
        "high": 4,
        "low": 2,
      },
      "prices": Object {
        "high": 2100,
        "low": 800,
      },
      "totalAvailable": 42,
    },
    "isActive": true,
    "isPaid": true,
    "listingId": 0,
    "originalListingProperty": "im available to children",
  },
  "someOtherProp": "im availabile to children",
}
`;
